File: dashboard/src/lib/__tests__/date-format.test.ts
// Moneat - observability platform
// Copyright (C) 2026 Moneat
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

import { describe, it, expect, beforeEach, afterEach } from 'vitest'
import {
  parseDate,
  formatDateTime,
  formatDate,
  formatMonthDay,
  formatTime,
  formatTimeHM,
  formatTimeHM12,
  formatDateTimeWithMs,
  formatTimeWithMs,
  formatDateTimeFull,
  browserTimezone,
} from '../date-format'

const ISO = '2025-01-15T14:34:56.789Z'
const DATE_OBJ = new Date(ISO)
const EPOCH_MS = DATE_OBJ.getTime()
const TZ = 'UTC'

// ──── parseDate ────

describe('parseDate', () => {
  let originalDate: typeof Date

  beforeEach(() => {
    originalDate = global.Date
  })

  afterEach(() => {
    global.Date = originalDate
  })

  it('handles ClickHouse DateTime64 format in strict mobile date parsers', () => {
    global.Date = class extends originalDate {
      constructor(value?: string | number | Date) {
        if (value === undefined) {
          super()
        } else if (typeof value === 'string' && value.includes(' ')) {
          super(Number.NaN)
        } else {
          super(value)
        }
      }
    } as unknown as DateConstructor

    expect(parseDate('2025-01-15 14:34:56.789').toISOString()).toBe(ISO)
  })
})

// ──── formatDateTime ────

describe('formatDateTime', () => {
  it('formats a Date object', () => {
    const result = formatDateTime(DATE_OBJ, TZ)
    expect(result).toContain('Jan')
    expect(result).toContain('15')
    expect(result).toContain('34')
    expect(result).toContain('56')
  })

  it('formats a string date', () => {
    const result = formatDateTime(ISO, TZ)
    expect(result).toContain('Jan')
    expect(result).toContain('15')
  })

  it('formats a numeric timestamp', () => {
    const result = formatDateTime(EPOCH_MS, TZ)
    expect(result).toContain('Jan')
    expect(result).toContain('15')
  })

  it('returns the raw string for an invalid date', () => {
    expect(formatDateTime('invalid', TZ)).toBe('invalid')
  })
})

// ──── formatDate ────

describe('formatDate', () => {
  it('formats a Date object', () => {
    const result = formatDate(DATE_OBJ, TZ)
    expect(result).toContain('Jan')
    expect(result).toContain('15')
    expect(result).toContain('2025')
  })

  it('formats a string date', () => {
    const result = formatDate(ISO, TZ)
    expect(result).toContain('2025')
  })

  it('formats a numeric timestamp', () => {
    const result = formatDate(EPOCH_MS, TZ)
    expect(result).toContain('2025')
  })

  it('returns the raw string for an invalid date', () => {
    expect(formatDate('invalid', TZ)).toBe('invalid')
  })
})

// ──── formatMonthDay ────

describe('formatMonthDay', () => {
  it('formats a Date object', () => {
    const result = formatMonthDay(DATE_OBJ, TZ)
    expect(result).toContain('Jan')
    expect(result).toContain('15')
  })

  it('formats a string date', () => {
    const result = formatMonthDay(ISO, TZ)
    expect(result).toContain('Jan')
  })

  it('formats a numeric timestamp', () => {
    const result = formatMonthDay(EPOCH_MS, TZ)
    expect(result).toContain('15')
  })

  it('returns the raw string for an invalid date', () => {
    expect(formatMonthDay('invalid', TZ)).toBe('invalid')
  })
})

// ──── formatTime ────

describe('formatTime', () => {
  it('formats a Date object in 24h', () => {
    const result = formatTime(DATE_OBJ, TZ)
    expect(result).toContain('14')
    expect(result).toContain('34')
    expect(result).toContain('56')
  })

  it('formats a string date', () => {
    const result = formatTime(ISO, TZ)
    expect(result).toContain('34')
  })

  it('formats a numeric timestamp', () => {
    const result = formatTime(EPOCH_MS, TZ)
    expect(result).toContain('56')
  })

  it('returns the raw string for an invalid date', () => {
    expect(formatTime('invalid', TZ)).toBe('invalid')
  })
})

// ──── formatTimeHM ────

describe('formatTimeHM', () => {
  it('formats a Date object in 24h HH:MM', () => {
    const result = formatTimeHM(DATE_OBJ, TZ)
    expect(result).toContain('14')
    expect(result).toContain('34')
    expect(result).not.toContain('56')
  })

  it('formats a string date', () => {
    const result = formatTimeHM(ISO, TZ)
    expect(result).toContain('14')
  })

  it('formats a numeric timestamp', () => {
    const result = formatTimeHM(EPOCH_MS, TZ)
    expect(result).toContain('34')
  })

  it('returns the raw string for an invalid date', () => {
    expect(formatTimeHM('invalid', TZ)).toBe('invalid')
  })
})

// ──── formatTimeHM12 ────

describe('formatTimeHM12', () => {
  it('formats a Date object in 12h', () => {
    const result = formatTimeHM12(DATE_OBJ, TZ)
    expect(result).toContain('2')
    expect(result).toContain('34')
    expect(result).toMatch(/PM/i)
  })

  it('formats a string date', () => {
    const result = formatTimeHM12(ISO, TZ)
    expect(result).toMatch(/PM/i)
  })

  it('formats a numeric timestamp', () => {
    const result = formatTimeHM12(EPOCH_MS, TZ)
    expect(result).toContain('34')
  })

  it('returns the raw string for an invalid date', () => {
    expect(formatTimeHM12('invalid', TZ)).toBe('invalid')
  })
})

// ──── formatDateTimeWithMs ────

describe('formatDateTimeWithMs', () => {
  it('includes milliseconds for a Date object', () => {
    const result = formatDateTimeWithMs(DATE_OBJ, TZ)
    expect(result).toContain('.789')
  })

  it('formats a string date with ms', () => {
    const result = formatDateTimeWithMs(ISO, TZ)
    expect(result).toContain('.789')
  })

  it('formats a numeric timestamp with ms', () => {
    const result = formatDateTimeWithMs(EPOCH_MS, TZ)
    expect(result).toContain('.789')
  })

  it('returns the raw string for an invalid date', () => {
    expect(formatDateTimeWithMs('invalid', TZ)).toBe('invalid')
  })
})

// ──── formatTimeWithMs ────

describe('formatTimeWithMs', () => {
  it('includes milliseconds for a Date object', () => {
    const result = formatTimeWithMs(DATE_OBJ, TZ)
    expect(result).toContain('14')
    expect(result).toContain('.789')
  })

  it('formats a string date', () => {
    const result = formatTimeWithMs(ISO, TZ)
    expect(result).toContain('.789')
  })

  it('formats a numeric timestamp', () => {
    const result = formatTimeWithMs(EPOCH_MS, TZ)
    expect(result).toContain('.789')
  })

  it('returns the raw string for an invalid date', () => {
    expect(formatTimeWithMs('invalid', TZ)).toBe('invalid')
  })
})

// ──── formatDateTimeFull ────

describe('formatDateTimeFull', () => {
  it('includes weekday and timezone for a Date object', () => {
    const result = formatDateTimeFull(DATE_OBJ, TZ)
    expect(result).toContain('Wed')
    expect(result).toContain('Jan')
    expect(result).toContain('2025')
    expect(result).toMatch(/UTC|Coordinated/)
  })

  it('formats a string date', () => {
    const result = formatDateTimeFull(ISO, TZ)
    expect(result).toContain('Wed')
  })

  it('formats a numeric timestamp', () => {
    const result = formatDateTimeFull(EPOCH_MS, TZ)
    expect(result).toContain('2025')
  })

  it('returns the raw string for an invalid date', () => {
    expect(formatDateTimeFull('invalid', TZ)).toBe('invalid')
  })
})

// ──── browserTimezone ────

describe('browserTimezone', () => {
  it('returns a non-empty string', () => {
    const tz = browserTimezone()
    expect(typeof tz).toBe('string')
    expect(tz.length).toBeGreaterThan(0)
  })
})

File: backend/src/test/kotlin/com/moneat/utils/DateFormatTest.kt
// Moneat - observability platform
// Copyright (C) 2026 Moneat
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

package com.moneat.utils

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException
import kotlin.test.Test
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertTrue
import kotlin.test.fail

/**
 * Tests to ensure consistent ISO 8601 date formatting across all API responses.
 *
 * CONTEXT: ClickHouse returns dates in format "YYYY-MM-DD HH:MM:SS" but JavaScript/frontend
 * expects ISO 8601 format "YYYY-MM-DDTHH:MM:SS.000Z" (with T separator and timezone).
 *
 * All ClickHouse queries that return dates to the frontend MUST use:
 * formatDateTime(timestamp_field, '%Y-%c-%dT%H:%i:%S.000Z')
 *
 * This test validates that date strings follow the correct format and can be parsed
 * by standard ISO 8601 parsers (both Java and JavaScript).
 */
class DateFormatTest {

    /**
     * Expected format: ISO 8601 with milliseconds and UTC timezone
     * Pattern: YYYY-MM-DDTHH:MM:SS.sssZ
     * Example: 2026-02-12T20:58:30.000Z
     */
    private val iso8601Pattern = DateTimeFormatter.ISO_INSTANT

    @Test
    fun `valid ISO 8601 dates should parse correctly`() {
        val validDates =
            listOf(
                "2026-02-12T20:58:30.000Z",
                "2024-01-15T09:58:31.000Z",
                "2023-12-31T23:59:59.999Z",
                "2024-01-01T00:00:00.000Z"
            )

        validDates.forEach { dateStr ->
            try {
                iso8601Pattern.parse(dateStr)
                // Success - date parsed correctly
            } catch (e: DateTimeParseException) {
                fail("Date '$dateStr' should parse as ISO 8601 but failed: ${e.message}")
            }
        }
    }

    @Test
    fun `ClickHouse default format should NOT be accepted`() {
        val clickhouseDates =
            listOf(
                "2026-02-12 20:58:30.000", // Missing T separator and Z
                "2026-02-12 20:58:30", // Space instead of T
                "2026-02-12T20:58:30", // Missing .000Z
            )

        clickhouseDates.forEach { dateStr ->
            assertFailsWith<DateTimeParseException>(
                message = "ClickHouse format '$dateStr' should NOT parse as ISO 8601"
            ) {
                iso8601Pattern.parse(dateStr)
            }
        }
    }

    @Test
    fun `API response JSON dates should be ISO 8601 compliant`() {
        // Simulate API responses that should contain properly formatted dates
        val issueResponse =
            """
            {
                "id": "test-issue-id",
                "firstSeen": "2026-02-12T20:58:30.000Z",
                "lastSeen": "2026-02-13T15:30:00.000Z"
            }
            """.trimIndent()

        val json = Json.parseToJsonElement(issueResponse).jsonObject
        val firstSeen = json["firstSeen"]?.jsonPrimitive?.content
        val lastSeen = json["lastSeen"]?.jsonPrimitive?.content

        assertNotNull(firstSeen, "firstSeen should exist")
        assertNotNull(lastSeen, "lastSeen should exist")

        // Should parse without exceptions
        try {
            iso8601Pattern.parse(firstSeen)
            iso8601Pattern.parse(lastSeen)
            // Success - both dates parsed correctly
        } catch (e: DateTimeParseException) {
            fail("Date fields should be ISO 8601 compliant but failed: ${e.message}")
        }
    }

    @Test
    fun `ClickHouse formatDateTime pattern should produce correct format`() {
        // Document the correct ClickHouse pattern
        val clickhousePattern = "%Y-%c-%dT%H:%i:%S.000Z"

        // This pattern should produce: 2026-02-12T20:58:30.000Z
        // %Y = 4-digit year
        // %c = month (1-12)
        // %d = day (01-31)
        // T = literal T
        // %H = hour (00-23)
        // %i = minutes (00-59)
        // %S = seconds (00-59)
        // .000Z = literal milliseconds and UTC timezone

        // Verify the pattern is documented
        assertTrue(
            clickhousePattern.contains("T"),
            "Pattern must include T separator"
        )
        assertTrue(
            clickhousePattern.endsWith("Z"),
            "Pattern must end with Z for UTC"
        )
        assertTrue(
            clickhousePattern.contains("%Y-%c-%d"),
            "Pattern must include date components"
        )
        assertTrue(
            clickhousePattern.contains("%H:%i:%S"),
            "Pattern must include time components"
        )
    }

    @Test
    fun `date format validation helper should identify invalid formats`() {
        val validDate = "2026-02-12T20:58:30.000Z"
        val invalidDate1 = "2026-02-12 20:58:30"
        val invalidDate2 = "Invalid Date"
        val invalidDate3 = ""

        assertTrue(isValidIso8601(validDate), "Valid ISO 8601 should pass")
        assertFalse(isValidIso8601(invalidDate1), "ClickHouse format should fail")
        assertFalse(isValidIso8601(invalidDate2), "Invalid string should fail")
        assertFalse(isValidIso8601(invalidDate3), "Empty string should fail")
    }

    /**
     * Helper function to validate ISO 8601 date strings.
     * This can be used in integration tests or validation logic.
     */
    private fun isValidIso8601(dateStr: String): Boolean {
        return try {
            if (dateStr.isBlank()) return false
            iso8601Pattern.parse(dateStr)
            // Additional check: must contain T separator and end with Z
            dateStr.contains("T") && dateStr.endsWith("Z")
        } catch (e: DateTimeParseException) {
            false
        }
    }

    @Test
    fun `JavaScript Date constructor compatibility check`() {
        // These dates should work with: new Date(dateString) in JavaScript
        val javaScriptCompatibleDates =
            listOf(
                "2026-02-12T20:58:30.000Z",
                "2024-01-15T09:58:31.000Z"
            )

        javaScriptCompatibleDates.forEach { dateStr ->
            // Verify format matches what JavaScript Date expects
            assertTrue(
                dateStr.matches(Regex("""\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z""")),
                "Date '$dateStr' should match JS-compatible pattern"
            )
        }
    }
}
